### PR TITLE
#4055 allow infer_signature to read pandas ExtensionDtype

### PR DESCRIPTION
**Pull request by caleboverman**
## What changes are proposed in this pull request?

Allows `infer_signature` to handle dataframes containing ExtensionDtype, addresses #4055 

## How is this patch tested?

Modified `tests/models/test_model_input_examples.py` to contain additional datatypes.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

`infer_signature` no longer fails when given a pandas dataframe containing an `ExtensionDtype`

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by caleboverman** [comment]
@tomasatdatabricks looks like the previous tests were expecting `infer_signature` to set a binary column type to string in the signature. Not sure if that was done for a reason and if we should set the signature type to binary now.

**Comment by tomasatdatabricks** [comment]
@caleboverman It's actually the other way around, the tests asserts that binary is inferred as binary but it was inferred as string in your case. It's because pd.api.types.infer_dtype(col) returns "mixed" instead of "binary". I used bytearray instead of bytes in the test. I think it should work for both - i.e. both bytearray and bytes should be inferred as binary. Can you maybe do the same check we do fro numpy array objects?

Otherwise the PR looks great, thanks for the contribution!

File: mlflow/types/utils.py [hunk 1]
```diff
@@ -135,6 +135,45 @@ def _infer_numpy_dtype(dtype: np.dtype) -> DataType:
...
+            if x is None:
+                return True
+            elif any(map(lambda c: isinstance(x, c), self.classes)):
+                self.seen_instances += 1
+                return True
+            else:
+                return False
+
+    is_binary_test = IsInstanceOrNone(bytes, bytearray)
+    if all(map(is_binary_test, col)) and is_binary_test.seen_instances > 0:
```
Comment (tomasatdatabricks) [inline mlflow/types/utils.py:157]: since this is the most expensive test, can we put it at the end?

**Review by tomasatdatabricks — COMMENTED** [review]
Looks good, just one small request, after that lgtm.

**Review by tomasatdatabricks — APPROVED** [review]
lgtm! Thanks for the contirbution @caleboverman!

I've decided to merge this as is since the last requested change was a minor and we have changes coming to schema enforcement as a result of adding tensor support.

**Pull request merged**